analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/CompoundAssignment.cs

```diff
@@ -18,18 +18,46 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
+using SonarAnalyzer.SymbolicExecution.Constraints;
+
 namespace SonarAnalyzer.SymbolicExecution.Roslyn.OperationProcessors;
 
 internal sealed class CompoundAssignment : SimpleProcessor<ICompoundAssignmentOperationWrapper>
 {
     protected override ICompoundAssignmentOperationWrapper Convert(IOperation operation) =>
         ICompoundAssignmentOperationWrapper.FromOperation(operation);
 
-    protected override ProgramState Process(SymbolicContext context, ICompoundAssignmentOperationWrapper assignment)
+    protected override ProgramState Process(SymbolicContext context, ICompoundAssignmentOperationWrapper assignment) =>
+        ProcessNumericalCompoundAssignment(context, assignment)
+        ?? LearnNotNullFromCompoundAssignment(context.State, assignment)
+        ?? context.State;
+
+    private ProgramState ProcessNumericalCompoundAssignment(SymbolicContext context, ICompoundAssignmentOperationWrapper assignment)
     {
-        var state = context.SetOperationValue(SymbolicValue.NotNull);
-        return assignment.Target.TrackedSymbol(state) is { } symbol
-            ? state.SetSymbolValue(symbol, SymbolicValue.NotNull)
-            : state;
+        var state = context.State;
+        var leftNumber = state[assignment.Target]?.Constraint<NumberConstraint>();
+        var rightNumber = state[assignment.Value]?.Constraint<NumberConstraint>();
+        if (leftNumber is { } && rightNumber is { } && Calculate(assignment.OperatorKind, leftNumber, rightNumber) is { } constraint)
+        {
+            state = state.SetSymbolConstraint(assignment.Target.TrackedSymbol(state), constraint);
```

Comment:
I cannot find a case where `(assignment.Target.TrackedSymbol(state)` is null and has a number constrain, so I don't check.

Reply:
I don't think we can create one right now but I would definitely check for TrackedSymbol regardless for the code to be future-proof.